mono/metadata/threadpool-ms-io.c

```diff
@@ -62,23 +68,42 @@ enum {
 	AIO_OP_LAST
 };
 
+typedef enum {
+	UPDATE_EMPTY = 0,
+	UPDATE_ADD,
+	UPDATE_REMOVE_SOCKET,
+	UPDATE_REMOVE_DOMAIN,
+} ThreadPoolIOUpdateType;
+
 typedef struct {
 	gint fd;
 	MonoSocketAsyncResult *sockares;
-} ThreadPoolIOUpdate;
+} ThreadPoolIOUpdate_Add;
```

Comment:
Having the queue simply allows it to have multiple updates at the same
time. This would also not reduce the concurrency as the call to
'wait' unlock the mutex, and give a chance to another thread waiting for
the lock in mono_threadpool_ms_io_add to access the protected section too.
And we already wait for the updates to be applied in the case of the remove
socket and domain, so this is not a new pattern.

On Tuesday, August 25, 2015, Rodrigo Kumpera <notifications@github.com>
wrote:

> In mono/metadata/threadpool-ms-io.c
> <https://github.com/mono/mono/pull/1969#discussion_r37920347>:
>
> >  typedef struct {
> >  	gint fd;
> >  	MonoSocketAsyncResult *sockares;
> > -} ThreadPoolIOUpdate;
> > +} ThreadPoolIOUpdate_Add;
>
> That would make the whole point of an update queue kind of pointless.
>
> I worked around by registering it as a conservative root. Not the right
> solution, but one that works.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mono/mono/pull/1969/files#r37920347>.
>


-- 
Ludovic Henry
Runtime Software Engineer
Xamarin, Inc.